### PR TITLE
implement language tags for site/community in db and api

### PR DESCRIPTION
**Pull request by Nutomic**
Check all the commit messages for a list of changes.

### CODE REVIEW

**Comment by Nutomic** [comment]
I think instead of `Vec<LanguageId>`, it would make sense to use some kind of `Set<LanguageId>`, as items dont need to be ordered, and set directly provides an `intersect()` method which we need. Problem is that a normal HashSet or BTreeSet needs more allocations and cpu work, so it would be necessary to pull in something like [tinyset](https://crates.io/crates/tinyset). Also, diesel only returns vec for db queries, it seems there is no way to get another collection without manual conversion. So probably not worth the effort.

**Comment by Nutomic** [comment]
This should be ready to merge. Still needs api tests (which require updated lemmy-js-client), and a lot of additional testing (during and after lemmy-ui implementation of this feature). Also need to test that db upgrades are working fine.

**Comment by Nutomic** [comment]
Last commit is broken, adding the list of languages to CommunityView seems tricky.

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:338]: It *might* be better to use `SiteLanguage::update`, which is the higher up path, and should run this anyway, but I leave that to you.

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:140]: These are extremely inefficient, (its doing a `select * where language_id in (0,1,2,3,4,.....)`) and you'd probably do better to use a join here to select the rows, then loop over them to do a delete. 

To do this disjoint in SQL, I think it'd be:

```sql
select cl.language_id, 
sl.language_id 
from community_language cl
left outer join site_language sl
on sl.language_id = cl.language_id
where sl.language_id is null; // Fairly sure this is right, but test it
```

If that SQL works, lmk and I can help you build the diesel query

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:130]: Diesel might not let you do join + delete, but they can let you do a left outer join and then loop over the results. See below.

File: crates/db_schema/src/impls/community.rs
Comment (dessalines) [inline crates/db_schema/src/impls/community.rs:104]: Looks good to me.

File: crates/db_views/src/site_view.rs
Comment (dessalines) [inline crates/db_views/src/site_view.rs:18]: Views should never have multiple queries, they should always be a single select.

The languages should come back from `GetSiteResponse`, probably next to `all_languages`, maybe `site_languages`.

File: crates/db_views/src/structs.rs
Comment (dessalines) [inline crates/db_views/src/structs.rs:109]: counts are allowed here because they come back with the join, but not languages, move those up to GetSiteResponse.

File: crates/db_views_actor/src/community_view.rs
Comment (dessalines) [inline crates/db_views_actor/src/community_view.rs:64]: Same here, move these to GetCommunityResponse. You can tell this is bad because you have this for read, but not for the query builder.

File: crates/db_views_actor/src/structs.rs
Comment (dessalines) [inline crates/db_views_actor/src/structs.rs:46]: Same, move this to the response.

File: crates/db_views_actor/src/community_view.rs
Comment (dessalines) [inline crates/db_views_actor/src/community_view.rs:26]: No vecs in tuples obviously.

File: crates/api_crud/src/post/create.rs
Comment (dessalines) [inline crates/api_crud/src/post/create.rs:210]: 3 queries, when this could be simplified to one using inner join to find the intersection set.

A note, `Language::read_undetermined` could just be a hard-coded LanguageId number rather than a read, since this data comes from a static migration.

File: crates/api_crud/src/comment/update.rs
Comment (dessalines) [inline crates/api_crud/src/comment/update.rs:83]: Might as well make a helper function for `Option<LanguageId`, since you use this code both here and in UpdatePost.

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:168]: This shouldn't be reading from All languages, but should only be reading from site languages, the subset.

File: crates/api_crud/src/community/update.rs
Comment (dessalines) [inline crates/api_crud/src/community/update.rs:64]: You can delete the subset check above, and just make sure CommunityLanguage::update is doing the subset in SQL by reading from SiteLanguages.

File: crates/api_crud/src/post/update.rs
Comment (dessalines) [inline crates/api_crud/src/post/update.rs:87]: Use helper function mentioned above.

File: migrations/2022-09-08-102358_site-and-community-languages/down.sql
Comment (Nutomic) [reply migrations/2022-09-08-102358_site-and-community-languages/down.sql:2]: diesel migration redo works fine. Problem here is that local_user_language table etc were added in an earlier migration. Maybe it would make sense to move the update function there?

File: scripts/test.sh
Comment (Nutomic) [reply scripts/test.sh:12]: was by accident. fixed

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs:61]: fixed

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs:338]: No this is specifically to test the limit_languages() function.

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs:136]: done

File: crates/db_views/src/site_view.rs
Comment (Nutomic) [reply crates/db_views/src/site_view.rs:18]: Done, named it `discussion_languages`.

File: crates/db_views_actor/src/structs.rs
Comment (Nutomic) [reply crates/db_views_actor/src/structs.rs:46]: This is pretty easy for `CommunityResponse`, but theres also this:
```
pub struct ListCommunitiesResponse {
  pub communities: Vec<CommunityView>,
}
```

I could change it like one of the following, but either option seems rather confusing.
```
pub struct ListCommunitiesResponse {
  pub communities: Vec<CommunityView>,
  pub discussion_languages: Vec<Vec<LanguageId>>
}
```

or 
```
pub struct ListCommunitiesResponse {
  pub communities: Vec<CommunityView, Vec<LanguageId>>,
}
```

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [inline crates/db_schema/src/impls/actor_language.rs:86]: With this transaction in place, many tests fail with error `Cannot perform this operation while a transaction is open`. But if i comment it out, tests pass just fine. Very strange.

File: crates/api_crud/src/comment/update.rs
Comment (Nutomic) [reply crates/api_crud/src/comment/update.rs:83]: done

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs:188]: Good catch. And im surprised that it doesnt break tests.

File: crates/api_crud/src/community/update.rs
Comment (Nutomic) [reply crates/api_crud/src/community/update.rs:64]: You mean as part of the insert statement? How would that work in sql?

File: less
Comment (dessalines) [inline less:1]: This got checked in wrongly.

File: migrations/2022-09-08-102358_site-and-community-languages/down.sql
Comment (dessalines) [reply migrations/2022-09-08-102358_site-and-community-languages/down.sql:2]: You need to undo those inserts you did in the up.sql (IE delete). Ideally every migration should be reversible.

File: crates/api_crud/src/post/create.rs
Comment (dessalines) [inline crates/api_crud/src/post/create.rs:206]: Still an issue, this intersect logic should be done in SQL.

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:152]: Still an issue.

File: crates/db_views_actor/src/structs.rs
Comment (dessalines) [reply crates/db_views_actor/src/structs.rs]: Why does the list of languages for each community need to come back when listing a lot of communities? Shouldn't only filtering be necessary?

File: crates/api_crud/src/community/update.rs
Comment (dessalines) [reply crates/api_crud/src/community/update.rs:64]: No, I mean inside the `CommunityLanguage::update(` fn, where it looks like you are already reading the site languages anyway. Looks like since users are providing any list of languages, you do have to read the site languages from sql, but can do a `.retain` type thing in that function.

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [reply crates/db_schema/src/impls/actor_language.rs]: Its trying to do tests at the same time then, weird. I thought that `#serial` marker was supposed to mitigate that.

File: less
Comment (Nutomic) [reply less]: fixed

File: crates/api_crud/src/post/create.rs
Comment (Nutomic) [reply crates/api_crud/src/post/create.rs]: I didnt finish addressing all the review comments yet, thats why they arent marked as resolved. You dont need to comment the same things again.

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs]: Maybe some tests are missing the marker.

File: crates/api_crud/src/community/update.rs
Comment (Nutomic) [reply crates/api_crud/src/community/update.rs:64]: I see, but i cant move the check in here because of LemmyError in return value. Then i would also have to return LemmyError from Community::create(), which is incompatible with the trait.

File: crates/api_crud/src/post/create.rs
Comment (Nutomic) [reply crates/api_crud/src/post/create.rs]: Fixed, and i was able to remove Language::read_undetermined entirely.

File: crates/db_views_actor/src/structs.rs
Comment (Nutomic) [reply crates/db_views_actor/src/structs.rs]: Ah, makes sense that languages are only needed in GetCommunityResponse. Added that.

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs]: There was a `#[serial]` missing in impls::person::tests::test_crud, but that didnt fix it. I suspect its because some apub tests use `#[actix_rt::test]` so its a different test runner and might run in parallel. Maybe we'll have to change all tests to use that annotiation. For now im leaving a todo here.

Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs]: Tried to fix it using block_on as in the link below, but didnt work.

https://stackoverflow.com/a/64597248

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs]: done

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs]: Found it, this calls `CommunityLanguage::limit_languages()` which also has a transaction, so there were two nested transactions. Removed the inner one now and it works.

File: migrations/2022-09-08-102358_site-and-community-languages/down.sql
Comment (Nutomic) [reply migrations/2022-09-08-102358_site-and-community-languages/down.sql:2]: So just delete everything in those tables? Wouldnt it be better in this case to move these update functions into the older migrations where tables local_user_language etc were created?

**Comment by Nutomic** [comment]
Okay made all the requested changes, except for the one in migrations (not sure whats the best solution), and the one in EditCommunity is not possible due to error types.

File: migrations/2022-09-08-102358_site-and-community-languages/down.sql
Comment (Nutomic) [reply migrations/2022-09-08-102358_site-and-community-languages/down.sql:2]: also done

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:69]: Put a TODO here because this subquery is bad. This should go away once we add a site:local column.

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:152]: Looks great, nice job.

File: crates/db_schema/src/impls/actor_language.rs
Comment (dessalines) [inline crates/db_schema/src/impls/actor_language.rs:209]: Perfect. I think I need some way to get this info to the front end tho.

**Review by dessalines — APPROVED** [review]
Nice job, the main thing I was concerned about was the intersection logic being in SQL. Feel free to merge once you add the TODO.

**Comment by dessalines** [comment]
Once this is merged, I'll start on the front end for it.

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs:210]: Added this to GetCommunityView.

File: crates/db_schema/src/impls/actor_language.rs
Comment (Nutomic) [reply crates/db_schema/src/impls/actor_language.rs:70]: done

**Pull request merged**
